src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala

```diff
@@ -211,13 +213,38 @@ abstract class BCodeIdiomatic {
       invokevirtual(JavaStringBuilderClassName, "append", bt.descriptor, pos)
     }
 
-    /*
+    /* Extract the built `String` from the `StringBuilder`
+     *:
      * can-multi-thread
      */
-    final def genEndConcat(pos: Position): Unit = {
+    final def genStringBuilderEnd(pos: Position): Unit = {
       invokevirtual(JavaStringBuilderClassName, "toString", "()Ljava/lang/String;", pos)
     }
 
+    /* Concatenate top N arguments on the stack with `StringConcatFactory#makeConcatWithConstants`
+     * (only works for JDK 9+)
+     *
+     * can-multi-thread
+     */
+    final def genIndyStringConcat(
+      recipe: String,
+      argTypes: Seq[asm.Type],
+      constants: Seq[String]
+    ): Unit = {
+      jmethod.visitInvokeDynamicInsn(
+        "makeConcatWithConstants",
+        asm.Type.getMethodDescriptor(StringRef.toASMType, argTypes:_*),
+        new asm.Handle(
+          asm.Opcodes.H_INVOKESTATIC,
+          "java/lang/invoke/StringConcatFactory",
```

Comment:
Since `java/lang/invoke/StringConcatFactory` is not necessarily present on the host JVM, the Scala compiler can't look up its name by reflection. I'd love some suggestions for what to do about this sort of constant.

Reply:
This is a good question. There is a contract in the backend that code generation instantiates a `ClassBType` for every type that appears in the bytecode. These [backend types are cached](https://github.com/scala/scala/blob/628d022291ab65e2e2b158497346de9b38a9b9c4/src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala#L56), and they are assumed to be present in the cache when [computing the stack map frames](https://github.com/scala/scala/blob/628d022291ab65e2e2b158497346de9b38a9b9c4/src/compiler/scala/tools/nsc/backend/jvm/PostProcessor.scala#L157), which happens during classfile writing.

However, here we're not pushing an object of type `StringConcatFactory` onto the stack, we're only using one of its static methods as an indy bootstrap method. So "it should be fine"™️.